### PR TITLE
Create the service scale up component

### PR DESCRIPTION
**Pull request by hatched**

### CODE REVIEW

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2951/
Test FAILed.

File: jujugui/static/gui/src/app/app.js [hunk 1]
```diff
@@ -769,6 +770,10 @@ YUI.add('juju-gui', function(Y) {
           metadata={metadata}>
           <components.Inspector
             service={service}
+            addGhostAndEcsUnits={utils.addGhostAndEcsUnits.bind(
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/app.js]: Nice way of handling this!

**Comment by huwshimi** [comment]
This seems like a fairly sensible way of approaching this. I don't have any big concerns.

File: jujugui/static/gui/src/app/components/env-size-display/env-size-display.js [hunk 2]
```diff
@@ -29,6 +29,11 @@ YUI.add('env-size-display', function() {
       };
     },
 
+    componentWillReceiveProps: function() {
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/env-size-display/env-size-display.js]: Document function?

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss [hunk 3]
```diff
@@ -0,0 +1,51 @@
+.scale-service {
+
+    input[type=text] {
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: inputs are missing the inset box-shadow: http://ubuntudesign.github.io/vanilla-framework/demo/#forms. This should probably be set system wide.

Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: Can you target a class name instead of an element in case the element changes in the future

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+        input[type=text] {
+            background-color: transparent;
+        }
+
+        span {
+            position: absolute;
+            right: 25px;
+            top: 30px;
+            color: $mid-grey;
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: Should be $warm-grey

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    input[type=text] {
+        margin: 20px;
+        height: 28px;
+    }
+
+    &__units {
+        position: relative;
+
+        input[type=text] {
+            background-color: transparent;
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: Why is this required? Isnt the unit element on white already?

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+    input[type=text] {
+        margin: 20px;
+        height: 28px;
+    }
+
+    &__units {
+        position: relative;
+
+        input[type=text] {
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: Can this element have `margin-bottom: 13px;` to bring the radio buttons in line with design.

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        }
+    }
+
+    &__selector {
+
+      input[type=radio] {
+        opacity: 1;
+        margin: 10px 0px 10px 15px;
+
+        &+label {
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: Should be `color: $cool-grey;`

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+            top: 30px;
+            color: $mid-grey;
+        }
+    }
+
+    &__selector {
+
+      input[type=radio] {
+        opacity: 1;
+        margin: 10px 0px 10px 15px;
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: `margin: 10px 0 6px 15px` to match design

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+      div:last-of-type {
+        margin-bottom: 30px;
+      }
+
+    }
+
+    &__constraints {
+
+        input[type=text] {
+            margin-top: 5px;
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: Should be `color: $cool-grey;`

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        label {
+            margin-left: 30px;
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: Add to match design
```
font-size: 14px;
margin-bottom: 1px;
color: $cool-grey;
```

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss [hunk 11]
```diff
@@ -0,0 +1,51 @@
+.scale-service {
+
+    input[type=text] {
+        margin: 20px;
+        height: 28px;
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: `margin-bottom: 10px;` please

File: jujugui/static/gui/src/app/components/scale-service/scale-service.js [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+    render: function() {
+      var buttons = [{
+        title: 'Confirm',
+        action: this._scaleUpService
+      }];
+
+      return (
+        <div className="scale-service">
+          <div className="scale-service__units">
+            <input type="text" name="num-units" onChange={this._updateState}/>
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/scale-service.js]: `type="number"`?

File: jujugui/static/gui/src/app/components/scale-service/scale-service.js [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+                name="placement" type="radio"
+                onChange={this._toggleConstraints}
+                defaultChecked={true}
+                id="manually-place-units" />
+              <label htmlFor="manually-place-units">Manually place</label>
+            </div>
+          </div>
+          <div className={this._generateClasses()}>
+            <label htmlFor="cpu-constraint">CPU (GHZ)</label>
+            <input type="text"
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/scale-service.js]: `type="number"`?

File: jujugui/static/gui/src/app/components/scale-service/scale-service.js [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+              id="cpu-constraint"
+              name="cpu-constraint"
+              onChange={this._updateState}/>
+            <label htmlFor="cores-constraint">Cores</label>
+            <input type="text"
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/scale-service.js]: `type="number"`?

File: jujugui/static/gui/src/app/components/scale-service/scale-service.js [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+              id="cores-constraint"
+              name="cores-constraint"
+              onChange={this._updateState} />
+            <label htmlFor="ram-constraint">Ram (MB)</label>
+            <input type="text"
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/scale-service.js]: `type="number"`?

File: jujugui/static/gui/src/app/components/scale-service/scale-service.js [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+              id="ram-constraint"
+              name="ram-constraint"
+              onChange={this._updateState} />
+            <label htmlFor="disk-constraint">Disk (MB)</label>
+            <input type="text"
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/scale-service.js]: `type="number"`?

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+
+      div:last-of-type {
+        margin-bottom: 30px;
+      }
+
+    }
+
+    &__constraints {
+
+        input[type=text] {
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: Can we target classes over raw elements unless its a style that is app wide otherwise these are component styles. For example if the someone changes this element to `type="number"` they will have to update the styles.

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+            margin-top: 5px;
+        }
+
+        label {
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: Same here, can we target a class please

File: jujugui/static/gui/src/app/components/scale-service/scale-service.js [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+          'root-disk': this.state['disk-constraint']
+        });
+      }
+      this.props.changeState(appState);
+    },
+
+    render: function() {
+      var buttons = [{
+        title: 'Confirm',
+        action: this._scaleUpService
```
Comment (anthonydillon) [inline jujugui/static/gui/src/app/components/scale-service/scale-service.js]: The design has Cancel. I'm not sure what it would do but just a note.

**Comment by anthonydillon** [comment]
Sorry @hatched a lot of comments but only small tweaks. QA OK.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2983/
Test PASSed.

File: jujugui/static/gui/src/app/app.js (hunk 1 above)
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: Thanks!

File: jujugui/static/gui/src/app/components/env-size-display/env-size-display.js (hunk 2 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/env-size-display/env-size-display.js]: I can do that however I haven't been documenting the native React functions as it's much better documented in the react docs.

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss (hunk 11 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: done

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss (hunk 6 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: done

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss (hunk 5 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: The background of the element is white I need it to be transparent so that the 'units' that is placed behind it is visible. This is required because 'placeholder' places it on the left.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2988/
Test PASSed.

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss (hunk 4 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: done

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss (hunk 8 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: I guesssed, looks like I was close ;)

done

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss (hunk 7 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: done

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss (hunk 10 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: done

File: jujugui/static/gui/src/app/components/scale-service/scale-service.js (hunk 19 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/scale-service/scale-service.js]: Yeah, but after a discussion with Luca we decided that it really didn't serve any purpose as it functioned exactly as the back button.

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss (hunk 3 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: Yes that should probably be set app wide so I won't tackle it here.

done converting to class.

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss (hunk 17 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: done - there ended up being some weird specificities to that we can get past some incorrectly applied global styles.

File: jujugui/static/gui/src/app/components/scale-service/_scale-service.scss (hunk 18 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/scale-service/_scale-service.scss]: done

File: jujugui/static/gui/src/app/components/scale-service/scale-service.js (hunk 12 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/scale-service/scale-service.js]: Can't change this to a number because then it puts the increment arrows on the right side of the inputs and modifies a number of other styles. We can revisit this later.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2989/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2991/
Test FAILed.

**Comment by hatched** [comment]
npm error

**Comment by hatched** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge
